include/bk/lexer.h

```diff
@@ -0,0 +1,54 @@
+/*
+   Copyright 2019 Simon (psyomn) Symeonidis
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+#pragma once
+
+#include <bk/token.h>
+#include <bk/module.h>
+
+#include <iostream>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+namespace bk {
+struct Lexer {
```

Comment:
Questions that you always need to ask yourself when creating a class:
- Should it be default-constructible (constructor with no arguments)?
- Can it be copied?
- Can it be moved?
The answer to these questions will influence which constructor and operators you should define yourself, delete or use the default one. 

Some reading material: https://cpppatterns.com/patterns/rule-of-five.html
Note that C++11 added a bit of syntax to automatically generate default implementations for most of these. An example:
```
struct Blah {
  Blah(const Blah&) = delete;
  Blah& operator= (const Blah&) = delete;
  Blah(Blah&&) = default;
  Blah& operator=(Blah&&) = default;
};
```
So in that example, I'm making an object that can't be copied (copy constructor and copy assignment operator are deleted) but can be moved (move constructor and move assignment operator are set to be auto-generated which boils down to calling the move operator on all the members of your class).

Reply:
Thank you!